### ISSUE TITLE
Keras with mxnet-backend and 8 GPUs trains slow when using ImageDataGenerator

### ISSUE THREAD

**Issue by Golbstein** [keras-team/keras#8028]
Following this great post: [https://devblogs.nvidia.com/parallelforall/scaling-keras-training-multiple-gpus/](url) I tried to upgrade my model to run in parallel on my multiple GPUs instance.
At first I ran the MNIST example as proposed here: [https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py](url) with the additional syntax in the compile command as follows:
```
# Prepare the list of GPUs to be used in training
NUM_GPU = 8 # or the number of GPUs available on your machine
gpu_list = []
for i in range(NUM_GPU): gpu_list.append('gpu(%d)' % i)

# Compile your model by setting the context to the list of GPUs to be used in training.
model.compile(loss='categorical_crossentropy',
              optimizer=opt,
              metrics=['accuracy'], 
              context=gpu_list)
```
then I trained the model:
```
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
```
So far so good. It ran for less than 1s per epoch and I was really excited and happy until I tried - **data augmentation.**

To that point, my training images were a numpy array at size (60000,1,28,28) and the labels were at size (10,60000) - one-hot encoded. For data augmentation I used the ImageDataGenerator function:

```
gen = image.ImageDataGenerator(rotation_range=8, width_shift_range=0.08, shear_range=0.3,
                               height_shift_range=0.08, zoom_range=0.08)
batches = gen.flow(x_train, y_train, batch_size=NUM_GPU*64)
test_batches = image.ImageDataGenerator().flow(x_test, y_test, batch_size=NUM_GPU*64)

model.fit_generator(batches, batches.N, nb_epoch=1, 
                    validation_data=test_batches, nb_val_samples=test_batches.N)
```
And unfortunately, from 1s per epoch I started getting ~11s per epoch... I suppose that the "impact" of the ImageDataGenerator is destructive and it probably running all the (reading->augmenting->writing to gpu) process really slow and inefficient.

Scaling keras to multiple GPUs is great, but data-augmentation is essential for my model to be robust enough.

Here are 2 benchmarks I made:

Using data augmentation:
![image](https://user-images.githubusercontent.com/31940058/31058777-fa434b44-a701-11e7-8c1e-20d6f14cd26a.png)

Without data-augmentation
![image](https://user-images.githubusercontent.com/31940058/31058771-e978fb6a-a701-11e7-8b03-4fdac8ef9955.png)

**Thanks!**

**Comment by JeniaNovellusDx** [comment]
Here are 2 benchmarks I made:

Using data augmentation:
![image](https://user-images.githubusercontent.com/31940058/31058258-837f6036-a6f9-11e7-9e33-37b781eb1cd3.png)

Without data-augmentation
![image](https://user-images.githubusercontent.com/31940058/31058266-9478145a-a6f9-11e7-8d96-0402a935d88b.png)

**Twice faster**

**Comment by unrealwill** [comment]
Hello, 
If you want to stay within keras you can probably do your data augmentation as a custom layer and put it into your model then push raw data.

Otherwise do the data loading directly in mxnet (you can probably use keras model to generate the mxnet graph). In mxnet you can then use a prefetchingIter (automatically in a thread).

**Comment by ivo-leonov** [comment]
Maybe using more threads to create the augmentations would speed things up a bit?
You could pass 'workers=N' to fit_generator, where N is the number of threads.

**Comment by DNXie** [comment]
I have the same problem. Before using `ImageDataGenerator` along with `fit_generator`, the code runs fine on multi gpus. When I tried these, it seems that the program runs only on one gpu, and it turns out to be very slow.
@ivo-leonov  I've tried what you said and it won't work. The program still runs on single gpu

**Comment by ivo-leonov** [comment]
@DNXie The 'workers' setting is only changing the number of CPU cores used for creating the augmentations. Nevertheless, weird behavior.
